fast64_internal/f3d/f3d_node_gen.py

```diff
@@ -0,0 +1,972 @@
+    "width_hidden": 42.0,
+    "height": 100.0,
+    "text": None,
+    "hide_value": False,
+    "subtype": "NONE",
+    "attribute_domain": "POINT",
+    "clamp": False,
+    "max_value": 3.4028234663852886e38,
+    "min_value": -3.4028234663852886e38,
+    "pin_gizmo": False,
+    "warning_propagation": "ALL",
+    "is_inspect_output": False,
+    "factor_mode": "UNIFORM",
+    "clamp_factor": True,
+    "clamp_result": False,
+    "use_alpha": False,
+    "use_clamp": False,
+    "normalize": True,
+    "blend_type": "MIX",
+    "noise_type": "FBM",
+    "mode": "RGB",
+    "enabled": False,  # this may seem like an odd default, but in versions with added inputs those would be set to False
+    # unused in shader nodes
+    "hide_in_modifier": False,
+    "force_non_field": False,
+    "layer_selection_field": False,
+}
+
+
+class DefaultDefinition:
+    def __init__(self, names: list[str], defaults: dict[str, Any]):
+        self.names = names
+        self.defaults = GLOBAL_DEFAULTS | defaults
+
+
+DEFAULTS = [
+    DefaultDefinition(["NodeSocketFloat", "NodeSocketFloatFactor"], {"default_value": 0.0}),
+    DefaultDefinition(["NodeSocketInt"], {"default_value": 0}),
+    DefaultDefinition(["NodeSocketVector", "NodeSocketRotation"], {"default_value": (0.0, 0.0, 0.0)}),
+    DefaultDefinition(["NodeSocketColor"], {"default_value": (0.0, 0.0, 0.0, 0.0)}),
+    DefaultDefinition(["ShaderNodeMixRGB"], {"data_type": "RGBA"}),
+    DefaultDefinition(["NodeSocketVectorDirection"], {"subtype": "DIRECTION", "default_value": (0.0, 0.0, 0.0)}),
+]
+DEFAULTS = {name: definition.defaults for definition in DEFAULTS for name in definition.names}
+
+SCENE_PROPERTIES_VERSION = 2
+
+
+class ErrorState:
+    def __init__(self, error_message_queue: list[str] = None):
+        self.error_message_queue = error_message_queue or []
+        self.errors = []
+
+    def copy(self, message: str = None):
+        errors = ErrorState(self.error_message_queue.copy() + ([message] if message else []))
+        errors.errors = self.errors
+        return errors
+
+
+def print_with_exc(error_state: ErrorState, exc: Exception):
+    message = "\n".join(error_state.error_message_queue)
+    print(message + ":\n" + str(exc))
+    print(traceback.format_exc())
+    error_state.errors.append((message, exc))
+
+
+def createOrUpdateSceneProperties():
+    group = bpy.data.node_groups.get("SceneProperties")
+    upgrade_group = bool(group and group.get("version", -1) < SCENE_PROPERTIES_VERSION)
+
+    if group and not upgrade_group:
+        # Group is ready and up to date
+        return
+
+    if upgrade_group and group:
+        # Need to upgrade; remove old outputs
+        if bpy.app.version >= (4, 0, 0):
+            for item in group.interface.items_tree:
+                if item.item_type == "SOCKET" and item.in_out == "OUTPUT":
+                    group.interface.remove(item)
+        else:
+            for out in group.outputs:
+                group.outputs.remove(out)
```

Comment:
iterating over sequences while modifying them is error-prone, I'd make copies of the sequences first
```suggestion
            for item in list(group.interface.items_tree):
                if item.item_type == "SOCKET" and item.in_out == "OUTPUT":
                    group.interface.remove(item)
        else:
            for out in list(group.outputs):
                group.outputs.remove(out)
```

Reply:
This is existing code from f3d_material.py so dont blame me :(